contrib/babelfishpg_tsql/runtime/functions.c

```diff
@@ -1727,6 +1730,204 @@ object_name(PG_FUNCTION_ARGS)
+    object_name = splitted_object_name[3];
+
+    /* downcase identifier if needed */
+    if (pltsql_case_insensitive_identifiers)
+    {
+        db_name = downcase_identifier(db_name, strlen(db_name), false, false);
+        schema_name = downcase_identifier(schema_name, strlen(schema_name), false, false);
+        object_name = downcase_identifier(object_name, strlen(object_name), false, false);
+        for (int i = 0; i < 4; i++)
+            pfree(splitted_object_name[i]);
+    }
+    else
+        pfree(splitted_object_name[0]);
+
+    pfree(input);
+    pfree(splitted_object_name);
+
+    /* truncate identifiers if needed */
+    truncate_tsql_identifier(db_name);
+    truncate_tsql_identifier(schema_name);
+    truncate_tsql_identifier(object_name);
+
+    if (!strcmp(schema_name, ""))
+    {
+        // To check if it is a system datatype, search in typecode list and it will give result oid, else if not it will return null.
+        result = (*common_utility_plugin_ptr->get_tsql_datatype_oid) (object_name);
+
+        // If null result, then it is not system datatype and now search with default schema in pg_type
+        if (!OidIsValid(result))
+        {
+            /* find the default schema for current user and get physical schema name */
+            const char *user = get_user_for_database(db_name);
+            const char *guest_role_name = get_guest_role_name(db_name);
+
+            if (!user)
+            {
+                pfree(db_name);
+                pfree(schema_name);
+                pfree(object_name);
+                PG_RETURN_NULL();
+            }
+            else if ((guest_role_name && strcmp(user, guest_role_name) == 0))
+            {
+                physical_schema_name = pstrdup(get_guest_schema_name(db_name));
+            }
+            else
+            {
+                pfree(schema_name);
+                schema_name = get_authid_user_ext_schema_name((const char *) db_name, user);
+                physical_schema_name = get_physical_schema_name(db_name, schema_name);
+            }
+        }
+        else
+        {
+            pfree(db_name);
+            pfree(schema_name);
+            pfree(object_name);
+            PG_RETURN_INT32(result);
+        }
+    }
+    else
+    {
+        physical_schema_name = get_physical_schema_name(db_name, schema_name);
+    }
+
+    /* get schema oid from physical schema name, it will return InvalidOid if user don't have lookup access */
+    schema_oid = get_namespace_oid(physical_schema_name, true);
+    pfree(db_name);
+    pfree(physical_schema_name);
+
+    if (!OidIsValid(schema_oid) || pg_namespace_aclcheck(schema_oid, user_id, ACL_USAGE) != ACLCHECK_OK)
+    {
+        pfree(schema_name);
+        pfree(object_name);
+        PG_RETURN_NULL();
+    }
+    // If schema is 'sys' or 'pg_catalog' then search in typecode list, else in pg_type.
+    if(!strcmp(schema_name, "sys") || !strcmp(schema_name, "pg_catalog"))
```

Comment:
Can we move this condition to before line 1850, we should save some execution in case of sys and pg_catalog